corehq/apps/reports/standard/deployments.py

```diff
@@ -216,8 +220,70 @@ def user_query(self, pagination=True):
                                        )
         return user_query
 
+    def get_location_columns(self, grouped_ancestor_locs):
+        from corehq.apps.locations.models import LocationType
+        location_types = LocationType.objects.by_domain(self.domain)
+        all_user_locations = grouped_ancestor_locs.values()
+        all_user_loc_types = {loc.location_type_id for user_locs in all_user_locations for loc in user_locs}
+        required_loc_columns = [loc_type for loc_type in location_types if loc_type.id in all_user_loc_types]
+        return required_loc_columns
+
+    def user_locations(self, ancestors, location_types):
+        ancestors_by_type_id = {loc.location_type_id: loc.name for loc in ancestors}
+        return [
+            ancestors_by_type_id.get(location_type.id, '---')
+            for location_type in location_types
+        ]
+
+    def get_bulk_ancestors(self, location_ids):
+        """
+        Returns the grouped ancestors for the location ids passed in the
+        dictionary of following pattern
+        {location_id_1: [self, parent, parent_of_parent,.,.,.,],
+        location_id_2: [self, parent, parent_of_parent,.,.,.,],
+
+        }
+        :param domain: domain for which locations is to be pulled out
+        :param location_ids: locations ids whose ancestors needs to be find
+        :param kwargs: extra parameters
+        :return: dict
+        """
+        where = Q(domain=self.domain, location_id__in=location_ids)
+        location_ancestors = SQLLocation.objects.get_ancestors(where)
+        location_by_id = {location.location_id: location for location in location_ancestors}
+        location_by_pk = {location.id: location for location in location_ancestors}
+
+        grouped_location = {}
+        for location_id in location_ids:
+
+            location_parents = []
+            current_location = location_by_id[location_id].id if location_id in location_by_id else None
+
+            while current_location is not None:
+                location_parents.append(location_by_pk[current_location])
+                current_location = location_by_pk[current_location].parent_id
+
+            grouped_location[location_id] = location_parents
+
+        return grouped_location
+
+    def include_location_data(self):
+        self.request.user = self.request.couch_user
```

Comment:
Oh yikes, yeah, I didn't notice that.  `request.user` is typically the Django user object, which is used for authentication and such.

Reply:
Because `self.request` is a FakeREquest Object not actual django request. and `enabled_for_request` function needs user attribute in the so Called `request` object passed to it. And since its a FakeRequest object it does not have `user` property but have `couch_user`

Reply:
It's a fake request?  How does that happen?  This looks like a normal report, and it frequently interacts with `self.request` as if it's a normal request, accessing a bunch of parameters from `self.request.GET`.  An alternative would be to do:
```python
toggle = toggles.LOCATION_COLUMNS_APP_STATUS_REPORT
return (
    toggle.enabled(username, toggles.NAMESPACE_USER)
    or toggle.enabled(domain, toggles.NAMESPACE_DOMAIN)
)
```
FWIW I actually don't like relying on `request` as much as we do, since it's a very complex, stateful object that's hard to reproduce and make guarantees about.

Reply:
hey, yeah its a fake http request. its created [here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/generic.py#L194-L228).  yeah, makes sense not touching which is not needed is better.  updating this check.

Reply:
I guess that's used whenever this is generated in celery?  Weird.

Another option here would be to switch `enabled_for_request` to check `request.couch_user`, which seems more appropriate anyways.  Looks like `TOGGLE.enabled` is called frequently with both types of users, though I think `couch_user` should always be used except for authentication.

I'm also happy with the solution in https://github.com/dimagi/commcare-hq/pull/24341/commits/77725b99e91837ea67425a256eac12e0b5dde385